### ISSUE TITLE
Problems with 4.0.8 / 4.0.9 and Traefik

### ISSUE THREAD

**Issue by freddydk** [microsoft/navcontainerhelper#2842]
We're experiencing the same issue.
The issue started appearing today.
We have three server hosting containers for our QA. They are all experiencing the same issue.
Containers created on friday are still up, running and reachable. 
We are running Docker Enterprise on Windows Server 2019  (1809)

<details>
<summary>docker ps shows "exactly" when the problem started occuring</summary>

```cmd
PS C:\Users\admjtheiner> docker ps -a
CONTAINER ID        IMAGE                                               COMMAND                  CREATED             STATUS                          PORTS
0b1064ca654d        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   2 minutes ago       Up About a minute (unhealthy)   80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
1717fa4da71e        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   5 hours ago         Up 5 hours (unhealthy)          80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
141ab654aa73        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   8 hours ago         Up 8 hours (unhealthy)          80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
b1cb4ed5be48        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   11 hours ago        Up 11 hours (unhealthy)         80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
72771723f0f5        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   12 hours ago        Up 12 hours (unhealthy)         80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
488ed96a322f        tobiasfenster/traefik-for-windows:v1.7.34           "/traefik --docker.e…"   21 hours ago        Up 21 hours                     0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp, 0.0.0.0:8080->8080/tcp
8238d719a686        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   4 days ago          Up 22 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
8a57b92b029b        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   4 days ago          Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
81b84a7ed5e3        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   7 days ago          Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
4375b693f606        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   11 days ago         Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
2415c1bafa62        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   4 weeks ago         Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
99ecd601ae19        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   4 weeks ago         Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
f43438446116        mcr.microsoft.com/businesscentral:10.0.17763.3650   "powershell -Command…"   5 weeks ago         Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
830228db2991        mcr.microsoft.com/businesscentral:10.0.17763.3406   "powershell -Command…"   3 months ago        Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
1009bef759f9        mcr.microsoft.com/businesscentral:10.0.17763.3287   "powershell -Command…"   4 months ago        Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
9e2fc9fa5129        mcr.microsoft.com/businesscentral:10.0.17763.2928   "powershell -Command…"   6 months ago        Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
cd0388a0843c        mcr.microsoft.com/businesscentral:10.0.17763.2366   "powershell -Command…"   11 months ago       Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
b3ec15f915c7        mcr.microsoft.com/businesscentral:10.0.17763.2366   "powershell -Command…"   12 months ago       Up 21 hours (healthy)           80/tcp, 443/tcp, 1433/tcp, 7045-7049/tcp, 7083/tcp, 8080/tcp
```

</details>
<details>
<summary>docker logs for the traefik container</summary>

```cmd
2023/01/08 23:34:17 Using high precision timer
2023/01/09 21:08:55 Using high precision timer
time="2023-01-09T21:08:57+01:00" level=error msg="Failed to retrieve information of the docker client and server host: error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running."
time="2023-01-09T21:08:57+01:00" level=error msg="Provider connection error error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running., retrying in 348.546422ms"
time="2023-01-09T21:08:57+01:00" level=error msg="Failed to retrieve information of the docker client and server host: error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running."
time="2023-01-09T21:08:57+01:00" level=error msg="Provider connection error error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running., retrying in 472.311552ms"
time="2023-01-09T21:08:58+01:00" level=error msg="Failed to retrieve information of the docker client and server host: error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running."
time="2023-01-09T21:08:58+01:00" level=error msg="Provider connection error error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running., retrying in 1.648796439s"
time="2023-01-09T21:08:59+01:00" level=error msg="Failed to retrieve information of the docker client and server host: error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running."
time="2023-01-09T21:08:59+01:00" level=error msg="Provider connection error error during connect: Get \"http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/version\": open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running., retrying in 855.729185ms"
```

</details>

We also recreate the traefik container each night. Since this image (1.7.34) is also 3 months old, I don't think it's the issue

I downgraded to v 4.0.7 of BcContainerHelper (not sure whether the issue appeared with .8 or .9) and recreated the container which didn't help. Then I recreated the traefik container and then recreated the container again, now everything seems to be working again. 

In future to make it easier to find what got released in which version of bccontainerhelper it would be very helpful if you could add tags when you release a new version so it's easier to compare releases.
I'm sorry I don't have time to look for what's actually going on but it must be a change between 4.0.7 and 4.0.9.

_Originally posted by @itsme112358 in https://github.com/microsoft/navcontainerhelper/issues/2839#issuecomment-1376389796_

**Comment by itsme112358** [comment]
Hi @freddydk,
thanks for migrating the issue.
Suggested Title: Traefik configuration does not update (BcContainerhelper 4.0.9)
Additional Info
The traefik configuration is not updated when a new container is created as it did before, rendering new containers unreachable when using `-useTraefik` parameter.
Existing containers work fine.
Scripts are run as local admin via a build agent with its own account.

Here is our traefik.toml 

```toml
debug = false
defaultEntryPoints = ["https","http"]
insecureSkipVerify = true

[api]
# Check https://docs.traefik.io/v1.7/configuration/api/#security
# to enable authentication on the dashboard for extra security

[docker]
domain = "[fqdn]"
watch = true
endpoint = "npipe:////./pipe/docker_engine"

[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
    entryPoint = "https"
  [entryPoints.https]
  address = ":443"
  [entryPoints.https.tls]
    [[entryPoints.https.tls.certificates]]
    certFile = "c:/etc/traefik/certificate.crt"
    keyFile = "c:/etc/traefik/certificate.key"
 
[file]
[backends]
  [backends.host]
    [backends.host.servers.server1]
       url = "http://${IP}:8180"

 [frontends]
  [frontends.host]
    backend = "host"
    [frontends.host.routes.route1]
      rule = "PathPrefix:/" 
```

And part of the calling script to create the traefik container
```powershell
$parameters = @{
    'IP' = $IPForTreafik;
    'PublicDnsName' = $env:DNSName;
    'CrtFile' = $CrtFile;
    'CrtKeyFile' = $CrtKeyFile;
    'traefikToml' = $traefikToml;
}

Setup-TraefikContainerForNavContainers -Recreate @parameters
```

Apart from timestamps and container id the only thing differing in the output is this
```cmd
2023-01-09T21:37:47.1950393Z DNS Name: [fqdn]
2023-01-09T21:37:47.9488603Z IP Address: [ipaddress]
2023-01-09T21:37:48.2140988Z BcContainerHelper version 4.0.7
2023-01-09T21:37:48.4744884Z BcContainerHelper emits usage statistics telemetry 

2023-01-08T22:41:29.5474994Z DNS Name: [fqdn]
2023-01-08T22:41:30.4890407Z IP Address: [ipaddress]
2023-01-08T22:41:30.7964664Z BcContainerHelper version 4.0.9
2023-01-08T22:41:31.1755032Z BC.HelperFunctions emits usage statistics telemetry to Microsoft
```

Both have been unchanged in ages.
We update BcContainerHelper nightly and recreate the traefik container.
Since over the weekend no builds ran, I can't say for sure whether the issue came up in 4.0.8 or 4.0.9

Since for now I can just work with 4.0.7, I'll stick with it for now. I have some deadlines I have to meet next week, so after that I'll be able to investigate further.

**Comment by freddydk** [comment]
Thanks for the additional information - I will investigate this

**Comment by itsme112358** [comment]
Hi Freddy, couldn't leave it and found the issue: The move to `Get-Acl` (09b364e3a10d270643f036e711ce513413c68395)
It works fine for "real" directories but it seems to have issues with the named pipe.
The commands I used to test this (as admin):
```powershell
$npipe = '//./pipe/docker_engine'

# works
[System.IO.Directory]::GetAccessControl($npipe)

Path Owner                        Access
---- -----                        ------
     VORDEFINIERT\Administratoren NT-AUTORITÄT\SYSTEM Allow  FullControl...

# doesn't work
Get-Acl -Path $npipe

# German, just states that the path could not be found because it doesn't exist.
Get-Acl : Der Pfad "Microsoft.PowerShell.Core\FileSystem:://./pipe/docker_engine" kann nicht gefunden werden, da er nicht vorhanden ist.
In Zeile:1 Zeichen:1
+ Get-Acl -Path $npipe
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (:) [Get-Acl], ItemNotFoundException
    + FullyQualifiedErrorId : GetAcl_PathNotFound,Microsoft.PowerShell.Commands.GetAclCommand
```
![image](https://user-images.githubusercontent.com/16080908/211585518-be326ff4-5f29-44c2-9b46-ff493874e45e.png)

**Comment by freddydk** [comment]
That was fixed in 4.0.9 - https://github.com/microsoft/navcontainerhelper/commit/d2916d91f9a5a930f371f60640365d4424955276
(which was the reason for shipping 4.0.9 very shortly after 4.0.8)

But I don't think it is the Check-BcContainerhelperPermission that fails with you

**Comment by freddydk** [comment]
I think I found the issue - testing now.
BTW - good point on the Tags - I will start adding tags to the repo on every release going forward.
(should have done that from day 1)

**Comment by freddydk** [comment]
Was able to easily find and add all the tags since 3.0.11 (all of these releases have been run on GitHub and the SHA was available for the bits released)

**Comment by freddydk** [comment]
Shipped in 4.0.10
